app/models/bot/keep.rb

```diff
@@ -54,29 +66,32 @@ def self.webhook(request)
     payload = JSON.parse(request.raw_post)
     if payload['url']
       link = Link.where(url: payload['url']).last
-      unless link.nil?
-        if payload['type'] == 'metrics'
-          Bot::Keep.update_metrics(link, payload['metrics'])
-        else
-          type = Bot::Keep.archiver_to_annotation_type(payload['type'])
-          response = Bot::Keep.set_response_based_on_pender_data(type, payload) || { error: true }
-          m = link.metadata_annotation
-          data = JSON.parse(m.get_field_value('metadata_value'))
-          data['archives'] ||= {}
-          data['archives'][payload['type']] = response
-          m.set_fields = { metadata_value: data.to_json }.to_json
-          m.save!
-
-          ProjectMedia.where(media_id: link.id).find_each do |pm|
-            annotation = pm.annotations.where(annotation_type: 'archiver').last
-
-            unless annotation.nil? || !DynamicAnnotation::Field.where(field_name: "#{type}_response", annotation_id: annotation.id).exists?
-              annotation = annotation.load
-              annotation.skip_check_ability = true
-              annotation.disable_es_callbacks = Rails.env.to_s == 'test'
-              annotation.set_fields = { "#{type}_response" => response.to_json }.to_json
-              annotation.save!
-            end
+      raise ObjectNotReadyError.new('link') unless link
+
+      if payload['type'] == 'metrics'
+        Bot::Keep.update_metrics(link, payload['metrics'])
+      else
+        type = Bot::Keep.archiver_to_annotation_type(payload['type'])
+        response = Bot::Keep.set_response_based_on_pender_data(type, payload) || { error: true }
+        m = link.metadata_annotation
+        data = JSON.parse(m.get_field_value('metadata_value'))
+        data['archives'] ||= {}
+        data['archives'][payload['type']] = response
+        m.set_fields = { metadata_value: data.to_json }.to_json
+        m.save!
+
+        project_media = ProjectMedia.where(media_id: link.id)
+        raise ObjectNotReadyError.new('project_media') unless project_media.count > 0
+
+        project_media.find_each do |pm|
+          annotation = pm.annotations.where(annotation_type: 'archiver').last
+
+          unless annotation.nil? || !DynamicAnnotation::Field.where(field_name: "#{type}_response", annotation_id: annotation.id).exists?
```

Comment:
here's the part I'm wondering about. what's the best way to know we should throw an exception?

Reply:
@hartsick , looking at the data we have, I think that in most cases it will be just one `ProjectMedia` per `Media`/`Link`, so I'm inclined to simplify based on that. That said, I think we can raise an exception when `Link` is `nil` or when at least one of the `ProjectMedia` doesn't have the `archiver` annotation. What do you think?